harness/determined/profiler.py

```diff
@@ -0,0 +1,524 @@
+        cpu_util_collector = SimpleCpuUtilCollector()
+        gpu_util_collector = GpuUtilCollector()
+        network_throughput_collector = NetThroughputCollector()
+        free_memory_collector = FreeMemoryCollector()
+        disk_collector = DiskReadWriteRateCollector()
+
+        while True:
+            if self.quitting:
+                # We drop any partial batches
+                break
+
+            if not self.is_active:
+                time.sleep(1)
+                continue
+
+            # One-time initialization
+            if last_measurement_time is None:
+                last_measurement_time = time.time()
+                batch_start_time = time.time()
+                network_throughput_collector.reset()
+                disk_collector.reset()
+
+
+            # Check if it is time to take a new measurement
+            if time.time() - last_measurement_time > self.MEASUREMENT_INTERVAL:
+                immutable_batch_idx = self.current_batch_idx
+                cpu_util_measurement = cpu_util_collector.measure(immutable_batch_idx)
+                gpu_util_measurements = gpu_util_collector.measure(immutable_batch_idx)
+                net_thru_sent_measurement, net_thru_recv_measurement = network_throughput_collector.measure(immutable_batch_idx)
+                free_memory_measurement = free_memory_collector.measure(immutable_batch_idx)
+                disk_read_thru_measurement, disk_write_thru_measurement, iops_measurement = disk_collector.measure(immutable_batch_idx)
+
+                for gpu_uuid in gpu_util_measurements.keys():
+                    self.current_batch.add_gpu_measurement(SysMetricType.GPU_UTIL_METRIC, gpu_uuid, gpu_util_measurements[gpu_uuid])
+
+                self.current_batch.add_nongpu_measurement(SysMetricType.NET_THRU_SENT_METRIC, net_thru_sent_measurement)
+                self.current_batch.add_nongpu_measurement(SysMetricType.NET_THRU_RECV_METRIC, net_thru_recv_measurement)
+                self.current_batch.add_nongpu_measurement(SysMetricType.DISK_IOPS_METRIC, iops_measurement)
+                self.current_batch.add_nongpu_measurement(SysMetricType.DISK_THRU_READ_METRIC, disk_read_thru_measurement)
+                self.current_batch.add_nongpu_measurement(SysMetricType.DISK_THRU_WRITE_METRIC, disk_write_thru_measurement)
+                self.current_batch.add_nongpu_measurement(SysMetricType.FREE_MEM_METRIC, free_memory_measurement)
+                self.current_batch.add_nongpu_measurement(SysMetricType.SIMPLE_CPU_UTIL_METRIC, cpu_util_measurement)
+
+                last_measurement_time = time.time()
+
+
+            # Check if it is time to flush the batch and start a new batch
+            if time.time() - batch_start_time > self.FLUSH_INTERVAL:
+                self.enqueue_for_async_send(self.current_batch.convert_to_post_format())
+                self.current_batch.clear()
+                batch_start_time = time.time()
+
+            time.sleep(0.02)
```

Comment:
I think it's probably just fine to sleep the (measurement_interval - time_taken) at the end of every run, that may even be more accurate since this could be off up to 0.02 + flush time if the condition on 302 is false and 325 is true on a pass through.